### ISSUE TITLE
Allow building releases

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#227]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/227)
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 5, 2018, 15:21

I don't think it's possible to actually make module releases using BuildStream. That is, there is no replacement for `jhbuild buildone -n --distcheck`. BuildStream should be able to run `make distcheck` for Autotools projects or `ninja dist` for Meson projects to generate a release tarball.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Feb 5, 2018, 16:34

FYI, this is a blocker to recommend GNOME developers to use buildstream instead jhbuild at the moment

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 5, 2018, 18:24

Summary of discussion with Mathieu in #buildstream:

I had been trying to use `bst shell`, but wasn't getting anywhere because I forgot about the need to use the `--build` flag. I was confused why the build directory was empty. Now, using `bst shell --build`, I can run `ninja dist` inside the build directory. But then, as Mathieu pointed out, there's no way to get the tarball outside the sandbox.

With hints from Mathieu, I eventually came up with this:

```$ bst shell --build core/epiphany.bst bash -- -c "cd _builddir; meson ..; ninja; ninja dist; cat epiphany-3.27.90.tar.xz > epiphany-3.27.90.tar.xz"```

Which is of course too difficult; we need a nicer way to do it. A syntax like `buildstream build --dist` would be ideal. Anyway, I'm not certain if that will work or not because the dist stage fails at:

```
In file included from ../tmp7c6a5wzi/epiphany-3.27.90/src/bookmarks/ephy-bookmark-properties-grid.h:25:0,
                 from src/ephy-type-builtins.c:5:
/usr/include/gtk-3.0/gtk/gtk.h:171:10: fatal error: /usr/include/gtk-3.0/gtk/gtkrecentchooser.h: Too many open files
 #include <gtk/gtkrecentchooser.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

which is insane. I guess it's running out of file descriptors?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 5, 2018, 18:30

To be able to access the generated release tarball, open a workspace for the target element before entering the build shell. The tarball will then be available in the workspace directory.

> I guess it's running out of file descriptors?

We will have to look into this. Maybe there is an issue with the FUSE layer.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 5, 2018, 18:38

> To be able to access the generated release tarball, open a workspace for the target element before entering the build shell. The tarball will then be available in the workspace directory.

Yes indeed, the tarball is actually already there in my workspace. I just hadn't noticed! That is good.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 5, 2018, 19:45

> A syntax like `buildstream build --dist` would be ideal.

This would still be nice to have, so I'll leave this issue open for BuildStream developers to decide if it should exist. But I no longer think this is urgent. Being able to generate a tarball in the workspace using `bst shell --build` is not hard, and should be plenty good enough for GNOME developers.

The fd issue is a blocker for me, though. Reported #232.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 6, 2018, 15:11

> The fd issue is a blocker for me, though. Reported #232.

Jürg found a workaround, which is `ninja dist -j1`, in order to disable parallel build. That's not stellar, but OK.

Next problem is that all of Epiphany's tests fail when generating the tarball. Running tests is not optional in Meson like it is in Autotools: there is no distinction between dist and distcheck. Most of the tests are failing because there is no running accessibility bus inside the sandbox. That's covered by #223, which seems hard to fix. Other tests are failing due to missing shared libraries, some of which I've never heard of (libbsd.so?). I don't understand why that is. Here's what I'm looking at:

```
[0/1] Running all tests.
 1/10 Completion model test                   FAIL     0.82 s
 2/10 Embed utils test                        FAIL     0.22 s
 3/10 Encodings test                          FAIL     0.22 s
 4/10 GSB service test                        FAIL     0.70 s
 5/10 History test                            FAIL     1.17 s
 6/10 Location entry test                     FAIL     1.05 s
 7/10 Migration test                          FAIL     0.60 s
 8/10 SQLite test                             FAIL     0.94 s
 9/10 String test                             FAIL     0.18 s
10/10 URI helpers test                        FAIL     0.18 s

OK:         0
FAIL:      10
SKIP:       0
TIMEOUT:    0


The output from the failed tests:

 1/10 Completion model test                   FAIL     0.82 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-completion-model
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-ephy-completion-model:3284): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 2/10 Embed utils test                        FAIL     0.22 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-embed-utils
--- stderr ---
/tmp/tmpkbmm0533/tests/test-ephy-embed-utils: error while loading shared libraries: libbsd.so.0: cannot open shared object file: Error 24
-------

 3/10 Encodings test                          FAIL     0.22 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-encodings
--- stderr ---
/tmp/tmpkbmm0533/tests/test-ephy-encodings: error while loading shared libraries: libbsd.so.0: cannot open shared object file: Error 24
-------

 4/10 GSB service test                        FAIL     0.70 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-gsb-service
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-ephy-gsb-service:3286): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 5/10 History test                            FAIL     1.17 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-history
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-ephy-history:3291): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 6/10 Location entry test                     FAIL     1.05 s

--- command ---
/tmp/tmpkbmm0533/tests/test-location-entry
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-location-entry:3289): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 7/10 Migration test                          FAIL     0.60 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-migration
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-ephy-migration:3288): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 8/10 SQLite test                             FAIL     0.94 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-sqlite
--- stderr ---

** (/tmp/tmpkbmm0533/tests/test-ephy-sqlite:3290): WARNING **: 15:05:29.120: Couldn't register with accessibility bus: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
-------

 9/10 String test                             FAIL     0.18 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-string
--- stderr ---
/tmp/tmpkbmm0533/tests/test-ephy-string: error while loading shared libraries: libICE.so.6: cannot open shared object file: Error 24
-------

10/10 URI helpers test                        FAIL     0.18 s

--- command ---
/tmp/tmpkbmm0533/tests/test-ephy-uri-helpers
--- stderr ---
/tmp/tmpkbmm0533/tests/test-ephy-uri-helpers: error while loading shared libraries: libsystemd.so.0: cannot open shared object file: Error 24
-------

Full log written to /tmp/tmpkbmm0533/meson-logs/testlog.txt
FAILED: meson-test 
/usr/bin/python3 -u /usr/local/bin/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.
Running unit tests on the distribution package failed.
FAILED: meson-dist 
/usr/bin/python3 /usr/local/bin/meson --internal dist /buildstream/build /buildstream/build/build /usr/bin/python3 /usr/local/bin/meson
ninja: build stopped: subcommand failed.
```

This makes me skeptical that tarball generation inside the BuildStream sandbox will be possible in the near future. I think we need to recommend going back to JHBuild, since we do need an immediate solution.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 6, 2018, 15:54

Turns out the failure to load shared libraries was also #232. So that's fixed. Now all the tests are failing due to missing accessibility bus. So that's progress.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 7, 2018, 15:42

D-Bus access seems to be impossible due to UID mismatch. Can you please test whether !265 solves this issue for you?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 7, 2018, 16:07

mentioned in merge request !265

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 7, 2018, 19:39

> D-Bus access seems to be impossible due to UID mismatch. Can you please test whether !265 solves this issue for you?

Yes, that does solve the issue! Cool.

That fixes half of Epiphany's tests. But the other half are now failing with `GLib-WARNING **: 19:37:06.386: getpwuid_r(): failed due to unknown user id (1000)` which is surely related to !265.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 7, 2018, 20:35

mentioned in commit 3b284b672017b7be09544de47321a3f8e63fe272

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 8, 2018, 11:12

A few notes on this while on vacation and hearing the echoes of disturbed screams of panicked individuals flailing arms about in hallways...

* Workspace approach sounds decent for now and fits the use case (i.e. it doesnt force distchecks on the unsuspecting, and just an interested party can generate one).
* Ultimately, like `make check`, I would prefer to wait for the caching of build trees feature to land and allow us to run things like `make check` or `make distcheck` in *separate* tasks later in the pipeline (i.e. `make check` fails a "build", but it does not block reverse dependencies from immediately starting to build alongside ongoing check tasks).
* Depending on host services already being alive at `make check` time is already wrong; I think EDS will at least pass it's tests because it takes care of creating and destroying the necessary services in it's test harness; modules which do not do this are literally running tests against whichever random version of the service happens to be active at the time, which seems clearly undesirable even in the case where you are running builds in an unprotected environment which lets this happen.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 13, 2018, 03:43

> Depending on host services already being alive at `make check` time is already wrong

Probably, yes

> I think EDS will at least pass it's tests because it takes care of creating and destroying the necessary services in it's test harness; modules which do not do this are literally running tests against whichever random version of the service happens to be active at the time, which seems clearly undesirable even in the case where you are running builds in an unprotected environment which lets this happen.

EDS runs *its own* services in its test harness. It does not set up an accessibility bus, which is required to run any test that uses GTK+. GTK+ unconditionally attempts to connect to the a11y bus after processing command line arguments and prints a warning (which will crash tests using GLib's test framework) if that fails. I don't think all GNOME modules should be individually responsible for figuring out how to start the a11y bus before running each test. See #223 where I suggest adding some sort of "GNOME environment" to set up certain D-Bus services inside the build sandbox.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 13, 2018, 13:05

> EDS runs its own services in its test harness. It does not set up an accessibility bus, which is required to run any test that uses GTK+. GTK+ unconditionally attempts to connect to the a11y bus after processing command line arguments and prints a warning (which will crash tests using GLib's test framework) if that fails.

This is a good point, however running tests against GTK+ has bigger problems than just this - and this is a problem we suffer a lot of in Glade as well, for most intents and purposes, it is impossible to use GTK+ in a command line tool because `gtk_init()` will fail unless it is able to connect to a display server of sorts (for glade, this prevents us from implementing interesting instrumentation tools which could validate a Glade file as a part of an automated build process, where a display server cannot be relied upon existing).

Almost anything which needs gtk_init() to work will break at `make check` time.

For #223, we certainly need some sort of project profile which can inform BuildStream of things like commands to run in the context of `bst shell` - but note that #223 is immensely different because we're talking about a shell environment which is intentionally weakened for the purpose of testing things out.

Running tests on the other hand is something you'd want to be entirely isolated like a build is; I'm personally more in favor of declaring that *"If you rely on system services at make check time, you're doing it wrong"* and go another step and say *"These should be installed tests"* - Further than this, it seems to me that installed tests need to grow some dependencies on the environment in which they mean to run.

Does an installed test want to run as root in a VM with no display server present yet ? Does an installed test want to run as a regular user ? Does it need to be a regular user which was logged in via GDM in a graphical session ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @TheRealMichaelCatanzaro]](https://gitlab.com/TheRealMichaelCatanzaro) on Feb 13, 2018, 13:42

> Does an installed test want to run as root in a VM with no display server present yet ? Does an installed test want to run as a regular user ? Does it need to be a regular user which was logged in via GDM in a graphical session ?

I don't know how our installed tests work, but I do know that the tests will probably never get run by developers unless they're run during `ninja dist` (or `make distcheck`). At least, that's the only time I ever run tests. Maybe with GitLab CI, installed tests would be more useful.

I've pushed https://git.gnome.org/browse/epiphany/commit/?id=3d674bbdf9713a2ea7bdeead6bee9e4c8e337b19 to epiphany for now, because I couldn't come up with any simple solution for this.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Feb 13, 2018, 14:26

> ... but note that #223 is immensely different because we're talking about a shell environment which is intentionally weakened for the purpose of testing things out

Replying to self here; the above as I understand it is incorrect; because we're using a `bst shell` environment in this scenario to generate release tarballs; so it already is running in a weakened sandbox environment.

Effort needs to be made in #223 to allow the project to configure things to run better (commands like `echo "8.8.8.8" > /etc/resolv.conf` for instance, could be run for an interactive `bst shell` session if the project could specify that, and also whitelisting of env vars which might be interesting to squash inside the shell).

In the long term however; I would like tests to run in a separate phase *inside the pipeline*; so that for instance; after building `glib`, we start building `gtk+` right away, while processing glib's test cases in parallel, any failed test can fail the build as a whole. This process could even be extended to allow creation of dist tarballs; but this will require more effort (depends on #21 which is currently underway) and would not support any communications with random things outside of sandboxes - in this scenario as well it would seem that customized commands to bring up services temporarily *inside* the sandbox would also be interesting.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 17, 2018, 06:59

mentioned in commit 7eb27f8f5c0e9f755206e850e03aeb3a229111d5

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 22, 2018, 13:47

mentioned in commit 11f2de77287ff06d60b88452aa125a37de375030

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 22, 2018, 16:54

closed via commit 11f2de77287ff06d60b88452aa125a37de375030
